### DISCUSSION TITLE
Error: Unable to find `next-intl` locale, have you configured the middleware?`

### DISCUSSION THREAD

**Discussion by 4Furki4** [amannn/next-intl#446] [Q&A] [answered]
Hi All !

I'm having a weird issue today that might be caused by next-intl. When I run my application both in the dev and prod build, I see this error in my console:

![image](https://github.com/amannn/next-intl/assets/84590614/007ec607-1083-4d0a-b11d-c3a96c929fb3)

The weird thing is that this error shows up only in my laptop and it's like an infinite loop! I have a team with 4 people, all of us tested to run the application and nothing is thrown and Vercel deployment is working actually.

-------------
Here's my middleware: 
![image](https://github.com/amannn/next-intl/assets/84590614/79041bc6-7466-4a99-a8aa-dac58344c1dd)

-------------
What I tried to solve it? 
- Deleting my project from local and re-clone it.
- recreating my middleware file which doesn't make sense at all bc I've never changed my middleware  

What might cause this error and how can I solve it ?

**Comment by RaenonX** [comment]
@amannn Pinging to bring up attention because I encountered the same issue.

What I discovered was that, there is actually some production request `http://localhost:3000/apple-touch-icon-120x120.png` happening. Somehow the matcher catches it, therefore giving this error.

The issue is that if I make the matcher `matcher: ['/((?!api|_next|(?:[.-]*)\\..*).*)'],` instead, `favicon.ico` will also be inaccessible for some reason.

Not sure if this got any solution for now unfortunately, I am out of ideas of changing the matcher so that these apple icons will be excluded.

**Reply by RaenonX** [reply]
@amannn Sorry for pinging again. Wanted to bring up attention because this is creating lots of false-negative in the error logs.

**Reply by amannn** [reply]
Without a reproduction, this is unfortunately not really actionable. If you can provide a minimal reproduction, please [open a bug report](https://github.com/amannn/next-intl/issues/new/choose) and I'll look into it. As you've probably seen, there's [a working example in the docs](https://next-intl-docs.vercel.app/examples/app-router) which can be used as a reference implementation when in doubt. Let me know how it goes!

**Reply by RaenonX** [reply]
Got it. I can't tell what went wrong yet but my app is showing a whole blank page for `/apple-touch-icon-120x120.png` while the sample app is showing the deault error page :/

Checking deeper tho

**Reply by RaenonX** [reply]
I think I actually used other version (could be older?) of the example, because the layout in the sample app you gave is so off from mine.

**Comment by lorenzoa7** [comment]
I'm getting the same error here.

**Comment by kjavier** [comment]
The working example provided uses NextIntlClientProvider. 

We are not using that and instead, we follow the one here: 
https://github.com/amannn/next-intl/blob/feat/next-13-rsc/

I think this version is the one causing us problems.

**Reply by pogran** [reply]
@kjavier this page https://github.com/amannn/next-intl/blob/feat/next-13-rsc/ doesn't available now. can you send right page? or mb are you find another reason how to solve this bug

**Reply by kjavier** [reply]
It is here https://github.com/amannn/next-intl/tree/feat/next-13-rsc.. Maybe we can do a try catch and redirect on the catch block.

**Reply by pogran** [reply]
i think need to find another module because its annoing error. its happend not usuall and only on prod 
![image](https://github.com/amannn/next-intl/assets/5382042/b9b201da-5556-4728-835e-4ffd896921c8)

**Reply by amannn** [reply]
If one of you can provide a minimal reproduction that shows a bug within next-intl, please feel free to open an issue and I'm happy to have a look! Based on the information available in this thread, it's unfortunately quite hard for me to understand where the issue is coming from.

The error message seems to show usage of `getLocale` in `generateMetadata`. Are you on the latest version of the `next-intl` RSC beta? `getLocale` was deprecated a few versions ago (you should see a deprecation hint on the console).

**Reply by pogran** [reply]
@amannn i use this version "next-intl": "^3.0.0-beta.14". I can set up this version which specified on docs https://next-intl-docs.vercel.app/docs/getting-started/app-router-server-components

**Reply by kjavier** [reply]
The problem is this file: https://github.com/amannn/next-intl/blob/feat/next-13-rsc/packages/next-intl/src/server/getLocaleFromHeader.tsx


For some reason, the cookies don't get set ALL THE TIME and this raises an error. 
```
if (!locale) {
    throw new Error(
      process.env.NODE_ENV !== 'production'
        ? 'Unable to find `next-intl` locale, have you configured the middleware?`'
        : undefined
    );
  }
```

Instead of raising an error, why not just return the default locale or get the cookies like this?
https://github.com/vercel/next.js/discussions/49444#discussioncomment-5866546


The issue stems from this problem in nextjs https://github.com/vercel/next.js/issues/49442.

**Reply by amannn** [reply]
> why not just return the default locale or get the cookies like this?

It could result in the wrong locale being used, which would be a bug. I think we should really get to the bottom of this.

> The issue stems from this problem in nextjs https://github.com/vercel/next.js/issues/49442.

In the issue the developer is using `response.cookies.set(…)`—notably the _response_. I don't think that's a bug. See also this comment: https://github.com/vercel/next.js/issues/49442#issuecomment-1679807704

In case the user doesn't have a cookie set yet (or it needs to be updated), we set a request header:

https://github.com/amannn/next-intl/blob/ab48fd2c0aca1b78fe1659c9039ef4bb07c70a7b/packages/next-intl/src/middleware/middleware.tsx#L128-L133

… which is then read during the RSC render:

https://github.com/amannn/next-intl/blob/ab48fd2c0aca1b78fe1659c9039ef4bb07c70a7b/packages/next-intl/src/server/getLocaleFromHeader.tsx#L11-L14

---

Since you seem to observe different behavior when running in development or when running the app in production (as far as I understand), there might be something else happening here.

If you can provide a minimal reproduction, please open an issue so I can investigate.

**Reply by kjavier** [reply]
The problem is it is hard to reproduce. I am not having the issue but a lot of our users are. And this is the same thing experienced by the main author of this discussion. He is the only one experiencing this error among his teammates. We could not determine how we can reproduce it. @4Furki4 are you able to determine the cause?

**Reply by amannn** [reply, marked as answer]
**Update Dec 12, 2023:** If you're still encountering this error, there's [a troubleshooting section in the docs now](https://next-intl-docs.vercel.app/docs/routing/middleware#unable-to-find-locale).

---

I've worked on a PR that could address this, please see the supplemental info here: https://github.com/amannn/next-intl/pull/506

Since this doesn't seem to have any downsides, I've released it as `next-intl@3.0.0-beta.17`. Can you have a look to see if it solves the issue you're seeing?

**Reply by kjavier** [reply]
Thanks! It seems like that would fix the issue. @pogran @RaenonX @lorenzoa7 @4Furki4  Would you be able to use next-intl@3.0.0-beta.17 on your applications? I still have to wait for our internal package repository to download the next-intl@3.0.0-beta.17 release before I can use it in our production. If you dont mind, please comment if the release fixes the issue.

**Reply by 4Furki4** [reply]
I've just updated the package to `next-intl@3.0.0-beta.17` and tested the middleware that was causing the issue on my side. It doesn't throw any errors and seems working seamlessly.
 
Thank you @amannn <3

**Reply by amannn** [reply]
🤝

**Reply by pogran** [reply]
@amannn in logs I have the same errors after update module to new version. my middleware

```
import createMiddleware from 'next-intl/middleware'
import { NextRequest, NextResponse } from 'next/server'

export default async function middleware(request: NextRequest) {
  const handleI18nRouting = createMiddleware({
    locales: ['ru'],
    defaultLocale: 'ru',
  })
  const response = handleI18nRouting(request)

  if (
    process.env.NEXT_PUBLIC_ENV_IS_MAINTENANCE === 'true' &&
    request.nextUrl.pathname !== '/maintenance'
  ) {
    return NextResponse.redirect(new URL('/maintenance', request.url), 302)
  }

  if (
    process.env.NEXT_PUBLIC_ENV_IS_MAINTENANCE !== 'true' &&
    request.nextUrl.pathname === '/maintenance'
  ) {
    return NextResponse.redirect(new URL('/', request.url), 302)
  }

  // https://nextjs.org/docs/pages/api-reference/functions/next-server#nextrequest

  // // Step 3: Alter the response

  return response
}

export const config = {
  // Skip all paths that should not be internationalized
  matcher: [
    '/((?!api|_next/static|_next/image|favicon.ico|apple-touch-icon.png|favicon.svg|images/books|icons|manifest).*)',
  ],
}
```

and logs 
![image](https://github.com/amannn/next-intl/assets/5382042/0c1b77f8-debd-4c21-b81c-f41ea90ed63e)

package.json 
![image](https://github.com/amannn/next-intl/assets/5382042/fe90fc27-a43a-488b-a7fe-4525ccba1e4c)

**Reply by pogran** [reply]
now i have more this type of error than before, I can try to reproduce this on my dev

**Reply by pogran** [reply]
and also i have question, why i see this error (**Unable to find `next-intl` locale, have you configured the middleware?**) in production if there we have this condition  https://github.com/amannn/next-intl/pull/506/files#diff-48a4e37781d0778726d6d368da10220c7e22ff4920d62b6b9f64bb7d780f18f5L34-L35 .

**Reply by amannn** [reply]
I think I'd need a minimal reproduction to continue here, otherwise it's too much guessing where the error comes from and where it can be fixed (next-intl, Next.js, your app code, deployment infrastructure, …).

**Reply by MatheoElieG** [reply]
Same issue here, sometimes when I run in dev no errors, but sometimes there are errors. I could try to see if on an other computer I get the same error then I can give you my repo url so you can try by yourself

**Reply by MatheoElieG** [reply]
Well, I just tried with an other computer (first time I had the problem was under ubuntu and now I tried on macos), the issue doesn't show up ...

**Reply by manelcomiche** [reply]
Same issue with `next-intl@3.0.0-beta.17`.

I attach the chunked code (just in case...)
![image](https://github.com/amannn/next-intl/assets/46402841/47a027bc-4e37-4447-b1d8-7b9c467d81c5)

**Comment by RaenonX** [comment]
I forgot what template I was using but using old/unupdated example is the
issue.

**Comment by kjavier** [comment]
Could it be the cause of this bug is this https://github.com/vercel/next.js/issues/49442

**Comment by MariuzM** [comment]
I'm on 3.0.0-beta.17 and `"next": "13.4.19",` still issue persists

**Comment by nicole0707** [comment]
You have to set up `X-NEXT-INTL-LOCALE` header in middleware and it will be consumed in server components.

**Reply by nicole0707** [reply]
Before we only set `NEXT_LOCALE` cookie, but there is a breaking change on `3.0.0-beta.17`.

**Reply by amannn** [reply]
If you use the required middleware, it shouldn't be breaking. Did you reimplement the middleware or was something else in your setup causing an issue?

**Reply by pogran** [reply]
@amannn is it possible to fork this version next-intl@3.0.0-beta.18? , i can't found how to do it . in tags i don't see this version, don't know how to find it

**Reply by MatheoElieG** [reply]
@amannn I'm using 3.0.0-beta.17

Here is my folder structure:

<img width="294" alt="image" src="https://github.com/amannn/next-intl/assets/113367146/6368f73c-c5f5-4efc-9d49-2c3b9096f809">

As you can see, it is very basic. This is my middleware:

```ts
import createMiddleware from 'next-intl/middleware';

export default createMiddleware({
    // A list of all locales that are supported
    locales: ['en', 'fr'],

    // If this locale is matched, pathnames work without a prefix (e.g. `/about`)
    defaultLocale: 'en',
});

export const config = {
    // Skip all paths that should not be internationalized. This example skips the
    // folders "api", "_next" and all files with an extension (e.g. favicon.ico)
    matcher: ['/((?!api|_next|.*\\..*).*)'],
};
```

All my translations are working correctly, which is good.

But I still have this error in my console:

<img width="632" alt="image" src="https://github.com/amannn/next-intl/assets/113367146/2cf75227-9e19-4c92-a3b4-17994ed234ac">

Here is my layout:

```ts
import '@styles/global.scss';
import { ReactNode } from 'react';
import { useLocale, useMessages, NextIntlClientProvider } from 'next-intl';
import { notFound } from 'next/navigation';
import { App_Roboto_Flex, App_Roboto_Mono } from '@libs/fonts';

interface LocaleLayoutProps {
    children: ReactNode;
    params: {
        locale: string;
    };
}

export default function LocaleLayout({ children, params }: LocaleLayoutProps) {
    const locale = useLocale();

    // Validate that the incoming `locale` parameter is a valid locale
    if (params.locale !== locale) {
        notFound();
    }

    const messages = useMessages();

    return (
        <html lang={locale}>
            <body className={`${App_Roboto_Flex.variable} ${App_Roboto_Mono.variable}`}>
                <NextIntlClientProvider messages={messages} locale={locale}>
                    {children}
                </NextIntlClientProvider>
            </body>
        </html>
    );
}
```

**Reply by amannn** [reply]
Hmm, I can’t spot an issue with the code you have provided. If you can reproduce this reliably, please open an issue and provide the necessary reproduction material. Thanks!

**Reply by MariuzM** [reply]
For me this issue was running as server component but when i switched to client all good

**Reply by MatheoElieG** [reply]
I tried to reproduce it but didn't succeed, sometimes I don't have error for 20 minutes, sometimes I keep getting errors.

**Reply by pogran** [reply]
imposible to find this bug... it reproduce in real project with big count of users. i also can't reproduce this in localhost, but in prod i have a lot of this errors

**Reply by manelcomiche** [reply]
I’ve been testing in prod too, with the same errors. I didn’t find anything.

**Reply by manelcomiche** [reply]
https://github.com/amannn/next-intl/discussions/446#discussioncomment-7026468

**Reply by amannn** [reply]
Maybe you can add logging and capture the whole request object in the middleware when the error occurs?

**Reply by nicole0707** [reply]
https://github.com/amannn/next-intl/blob/feat/next-13-rsc/packages/next-intl/src/server/getLocaleFromHeader.tsx#L9 It checks `X-NEXT-INTL-LOCALE` header now, if header isn't found, it will complain `Unable to find next-intl locale, have you configured the middleware?`

**Comment by pogran** [comment]
> You have to set up `X-NEXT-INTL-LOCALE` header in middleware and it will be consumed in server components.

i need to set up this x-next-intl-locale with nginx?

**Reply by nicole0707** [reply]
We add the header in Next Middleware, because we use Middleware chain and have to manually check and add header for some scenarios.

**Comment by scerelli** [comment]
I have the same error, i don't understand what should I do. should it work without any workaround in beta 17? or am I missing something?

This is the log of my middleware

```
middleware log: {
lms:dev:   cookies: ResponseCookies {},
lms:dev:   url: '',
lms:dev:   body: null,
lms:dev:   bodyUsed: false,
lms:dev:   headers: {
lms:dev:   x-middleware-override-headers: 'accept,accept-language,connection,cookie,dnt,host,referer,sec-fetch-dest,sec-fetch-mode,sec-fetch-site,upgrade-insecure-requests,user-agent,x-forwarded-for,x-forwarded-host,x-forwarded-port,x-forwarded-proto,x-invoke-path,x-invoke-query,x-middleware-invoke,x-next-intl-locale',
lms:dev:   x-middleware-request-accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8',
lms:dev:   x-middleware-request-accept-language: 'en-US,en;q=0.5',
lms:dev:   x-middleware-request-connection: 'close',
lms:dev:   x-middleware-request-cookie: 'sb-...; NEXT_LOCALE=it',
lms:dev:   x-middleware-request-dnt: '1',
lms:dev:   x-middleware-request-host: 'localhost:3000',
lms:dev:   x-middleware-request-referer: 'http://localhost:3000/courses',
lms:dev:   x-middleware-request-sec-fetch-dest: 'document',
lms:dev:   x-middleware-request-sec-fetch-mode: 'navigate',
lms:dev:   x-middleware-request-sec-fetch-site: 'same-origin',
lms:dev:   x-middleware-request-upgrade-insecure-requests: '1',
lms:dev:   x-middleware-request-user-agent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Gecko/20100101 Firefox/117.0',
lms:dev:   x-middleware-request-x-forwarded-for: '::ffff:127.0.0.1',
lms:dev:   x-middleware-request-x-forwarded-host: 'localhost:3000',
lms:dev:   x-middleware-request-x-forwarded-port: '3000',
lms:dev:   x-middleware-request-x-forwarded-proto: 'http',
lms:dev:   x-middleware-request-x-invoke-path: '',
lms:dev:   x-middleware-request-x-invoke-query: '',
lms:dev:   x-middleware-request-x-middleware-invoke: '1',
lms:dev:   x-middleware-request-x-next-intl-locale: 'it',
lms:dev:   x-middleware-rewrite: 'http://localhost:3000/it/courses'
lms:dev: },
lms:dev:   ok: true,
lms:dev:   redirected: false,
lms:dev:   status: 200,
lms:dev:   statusText: '',
lms:dev:   type: 'default'
lms:dev: }
lms:dev: - error ../node_modules/.pnpm/next-intl@3.0.0-beta.17_next@13.4.7_react@18.2.0/node_modules/next-intl/dist/esm/server/getLocaleFromHeader.js (1:578) @ eval
lms:dev: - error Error: Unable to find `next-intl` locale, have you configured the middleware?`
lms:dev:     at RootLayout (./src/app/[locale]/layout.tsx:36:73)
lms:dev:     at async Promise.all (index 0)
```


I also tried this:

```
import createMiddleware from 'next-intl/middleware'
import type { NextRequest } from 'next/server'
export default async function middleware(request: NextRequest) {
  const handleI18nRouting = createMiddleware({
    locales: ['it'],
    defaultLocale: 'it',
  })

  const response = handleI18nRouting(request)
  console.log('middleware log:', response)
  response.headers.set('X-NEXT-INTL-LOCALE', 'it')
  return response
}

export const config = {
  // Skip all paths that should not be internationalized
  matcher: ['/((?!api|_next|.*\\..*).*)']
}

```

same stuff as someone was suggesting but same story

**Comment by pogran** [comment]
i drop this module (temporary) from my project and it starts to work 2x faster. I think module have big problem with performance if you have some audience in project

**Reply by amannn** [reply]
Would be curious to hear what kind of performance issues you're seeing in your app.

https://next-intl-example-next-13.vercel.app/ currently has a PSI performance score of 100 from what I can see: https://pagespeed.web.dev/analysis/https-next-intl-example-next-13-vercel-app/qhsh1myjli?form_factor=mobile

<img width="956" alt="Screenshot 2023-09-19 at 15 52 43" src="https://github.com/amannn/next-intl/assets/4038316/b2b92763-c557-45f7-921e-02ca892b055f">


The adapted example in the RSC branch has the same score: https://pagespeed.web.dev/analysis/https-next-intl-example-next-13-git-feat-next-13-rsc-next-intl-vercel-app/powwa1foek?form_factor=mobile

<img width="955" alt="Screenshot 2023-09-19 at 15 52 34" src="https://github.com/amannn/next-intl/assets/4038316/7f45018f-c9c6-48f2-9b0c-137b64938397">

**Reply by pogran** [reply]
@amannn  yes, i think this connect with this error in this disscussion and workload. I also have this errors and mb that why i have some trobles with my site (but its fact that my users told me that site now works much better). That also fact i have 1-4 errors per 30sec - 1min. I think this error can be found with workload because i also can't reproduce it in my localhost but in prod i have a lot of this errors

**Reply by amannn** [reply]
Would be interesting if you could reproduce the bug with an HTTP load generator like [oha](https://github.com/hatoo/oha).

My results for `example-next-13` in the RSC branch locally (uses dynamic rendering currently):

<img width="504" alt="Screenshot 2023-09-20 at 12 58 11" src="https://github.com/amannn/next-intl/assets/4038316/415102f8-f468-464a-8841-92683bccb185">

After an upgrade to Next.js 13.5.1:

<img width="563" alt="Screenshot 2023-09-20 at 12 58 38" src="https://github.com/amannn/next-intl/assets/4038316/a9af5066-3624-4557-b970-38dc27187c41">

Next.js 13.5.1 generally includes many bug fixes, maybe this improves your situation too!

**Reply by amannn** [reply]
On top of that, there's now support for static rendering in the RSC integration, maybe that helps to reduce the load on your server too: [static rendering docs](https://next-intl-docs.vercel.app/docs/getting-started/app-router-server-components#static-rendering).

Benchmark results with static rendering:

<img width="511" alt="Screenshot 2023-09-22 at 09 22 03" src="https://github.com/amannn/next-intl/assets/4038316/3c150290-f4fc-4a2a-b4b8-4e521588b44f">

**Comment by manelcomiche** [comment]
I have managed to find out where these errors come from and why they only appear in production!

It is because of the 404 page. The pages not having 404 specifically created, it loads the one that has next by default, and that... makes the function jump.

It comes out in production because we constantly have crawlers that try to access wp-admins...

@amannn

**Reply by MariuzM** [reply]
can you try latest next version, i think they dont have 404 in pages dir anymore

**Comment by MatheoElieG** [comment]
I managed to fix the error in my console since we do not have a better solution yet and it was getting me cary to see all thoes errors.

This is my `src/app/[locale]/layout.tsx`:

```ts
import { ReactNode } from 'react';
import { notFound } from 'next/navigation';
import { headers } from 'next/headers';
import { useLocale, useMessages } from 'next-intl';

interface LocaleLayoutProps {
    children: ReactNode;
    params: {
        locale: string;
    };
}

export default function LocaleLayout({ children, params }: LocaleLayoutProps) {
    const header = headers();
    const localeHeader = header.get('x-next-intl-locale');
    if (localeHeader === null) {
        notFound();
    }

    const locale = useLocale();
    const messages = useMessages();

    // Validate that the incoming `locale` parameter is a valid locale
    if (params.locale !== locale) {
        notFound();
    }

    // ... rest of the code

}
```

Translations just works the same, but on the same request header without `x-next-intl-locale`somehow reaches the layout which should not happened since I have the middleware.

For debug purpose here is my headers list for a "healthy" request and an "unhealthy" request:

`healthy (with x-next-intl-locale)`:

```
host localhost:3000
connection keep-alive
sec-ch-ua "Chromium";v="116", "Not)A;Brand";v="24", "Google Chrome";v="116"
sec-ch-ua-mobile ?0
user-agent Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36
next-url /
sec-ch-ua-platform "macOS"
accept */*
sec-fetch-site same-origin
sec-fetch-mode cors
sec-fetch-dest empty
referer http://localhost:3000/
accept-encoding gzip, deflate, br
accept-language fr-FR,fr;q=0.9,en-US;q=0.8,en;q=0.7
cookie next-auth.csrf-token=b9344ee4ceae2e7ffecce7efbae4f8227b47149bf3fac7487cbc1f02be1f9a11%7Cddc02b08ce377ee779adfb8a18ede6b553a75658034df525206550f65962a1a0; NEXT_LOCALE=en; next-auth.callback-url=http%3A%2F%2Flocalhost%3A3000%2F%3Fauth-dialog%3Dsignin; next-auth.session-token=eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..KRO8CB0FouFZeO_A.kUgIWlrZRrulTp6_I7vtIKIS2RJ0BNM1iaXWZv5kNvgmsMkFrzD_3MLekiKLQXl857KnB_t1GLamJqoRF1UZOrOWQBMfG9SJ-IpMUUS0BY8AsRVSAs-xXlNQIDkJJWKIX7AhMLgkJt9k7rR-W7GEeOkqeHZMwLaZByA0PM8onwth-WIWYEFj46S93nUm6d0hW34MEBtgDJuX0v1sN4iStCy5BaZfu1xLI-NGqxGKgNQJ0cHACJjscFFFthpB0ee2UODBVngSe-Do1efLsrhfBnpK4uUd_W6fswzloc3EI-E_j_dwLzMsKZZaJQywKBDUz6bqAWF-U94.uzU55gofEOKJ7_HCCelBFw
x-next-intl-locale en
link <http://localhost:3000/>; rel="alternate"; hreflang="en", <http://localhost:3000/fr>; rel="alternate"; hreflang="fr", <http://localhost:3000/>; rel="alternate"; hreflang="x-default"
```

`unhealthy`:

```
host localhost:3000
connection keep-alive
pragma no-cache
cache-control no-cache
sec-ch-ua "Chromium";v="116", "Not)A;Brand";v="24", "Google Chrome";v="116"
sec-ch-ua-mobile ?0
user-agent Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36
sec-ch-ua-platform "macOS"
accept image/avif,image/webp,image/apng,image/svg+xml,image/*,*/*;q=0.8
sec-fetch-site same-origin
sec-fetch-mode no-cors
sec-fetch-dest image
referer http://localhost:3000/fr
accept-encoding gzip, deflate, br
accept-language fr-FR,fr;q=0.9,en-US;q=0.8,en;q=0.7
cookie next-auth.csrf-token=b9344ee4ceae2e7ffecce7efbae4f8227b47149bf3fac7487cbc1f02be1f9a11%7Cddc02b08ce377ee779adfb8a18ede6b553a75658034df525206550f65962a1a0; next-auth.callback-url=http%3A%2F%2Flocalhost%3A3000%2F%3Fauth-dialog%3Dsignin; NEXT_LOCALE=fr; next-auth.session-token=eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..oOqccYa_nIGzaZOU.l5F74eL_mUeW0qMwszFRTNufN_FCC79KLkBg2wz2xphrWgqUfprBQNgEyfTO8nEklWgSoDuInCuQ9JJ6-3IysQ4-XJdG5i3-oYzeUEyO5MlsCAoDX1jmY15rGQsqL8YAkCIX_cSvkKZ8theUfVr0V-pIggMpHnWYcQfdF-uZF_kV4mOXZ9r3KRgpIIDShB3CXfQ00xjtaNIyeZWOB4pPFf1cMR780R5-HZc9EDAj9UWrKGoKVJghXJBD4Mfh9B8TbZJQQDFPDZBeWSCa1Eq9xrCEicaRu8_K3vOGBYsCVzgsr6O0b5MuHWiBkEJoOrA0hZreLA2rMpA.nouOqkA6VHMuC13fS_3a1A
```

I tried to disable my extensions, nothing changed.
I feel like it only happens when I reload by pressing ctrl f5 (clearing cache)
Maybe header is not correctly populated when cache has just been cleared ?

**Reply by amannn** [reply]
Thanks for sharing the details! Based on your logs, the "unhealthy" request looks for an image (given the `accept` header). Have you configured the middleware to ignore images? In that case, if the image doesn't exist, the request will end up with an image file being the `locale` param (e.g. `image.png`).

See also [the matcher config docs](https://next-intl-docs.vercel.app/docs/routing/middleware#matcher-config).

**Reply by MatheoElieG** [reply]
Yes my middleware does ignore images since they are located under `_next/....`
This is my middleware:

```ts
import createMiddleware from 'next-intl/middleware';

export default createMiddleware({
    // A list of all locales that are supported
    locales: ['en', 'fr'],

    // If this locale is matched, pathnames work without a prefix (e.g. `/about`)
    defaultLocale: 'en',
});

export const config = {
    // Skip all paths that should not be internationalized. This example skips the
    // folders "api", "_next" and all files with an extension (e.g. favicon.ico)
    matcher: ['/((?!api|_next|.*\\..*).*)'],
};
```

**Reply by amannn** [reply]
Hmm, I see. Do you have a URL for the "unhealthy" request above? I don't find a "next-url" there.

**Reply by MatheoElieG** [reply]
Ok thank you for your time and I'm sorry for bothering you.
I changed my layout to output the params.locale in case of error, and as you just mentioned in the documentation, when the error occurs, my `params.locale` is equal to `favicon.ico`. But why does it happens since I just the recommended matcher that excluse all files with dots ? 

```ts
export const config = {
    // Skip all paths that should not be internationalized. This example skips the
    // folders "api", "_next" and all files with an extension (e.g. favicon.ico)
    matcher: ['/((?!api|_next|_vercel|.*\\..*).*)'],
};
```

Edit; I found out that I deleted the default `favicon.ico` from the default nextjs template. I tried to add it again, and the error disapeared. I guess it as something to do with the notFound error when I delete it, but I don't understand why nextjs try to find it since I don't have anything related to it in my code

**Comment by amannn** [comment]
There's a new troubleshooting section in the docs in regard to this error: https://next-intl-docs.vercel.app/docs/routing/middleware#unable-to-find-locale.

It will also be linked to from the "Unable to find next-intl locale" message in the next beta release.

**Comment by truong0vanchien** [comment]
As far as I know, this cause when you have a dot sign "." in your url string, the request will not go to middleware where we set next-intl locale. I was facing the same error, then I just manually change "." to "_DOT_" which is quite more friendly for a url string.

**Comment by jovand3v** [comment]
If anyone is still having this issue check if you've put the middleware file in the right place. As stated in the [Next.js docs](https://nextjs.org/docs/app/building-your-application/routing/middleware#convention).

> Use the file middleware.ts (or .js) in the root of your project to define Middleware. For example, at the same level as pages or app, or inside src if applicable.

**Comment by FilFarEu** [comment]
I was figuring this out for 3 hours.. my problem was that in middleware.js I had:

```
export const config = {
  matcher: ['/((?!api|_next|_vercel|.*\\..*).*)']
};
```

and I had to change it to:
```
export const config = {
  matcher: ['/((?!api|_next|_vercel\\..*).*)']
};
```

My problem was that I wanted to create 404 error file to customize it by myself..
http://localhost/randomstring - was working and throwing 404
http://localhost/random.string - was throwing big error and page was not visible at all

So only "." - dot I needed to remove to actually catch it right.

This was only my case.

**Reply by MichalMoravik** [reply]
Changing your matcher according to this comment will malfunction any place where you use static files straight from your `/public` folder. 

As I use quite a few of these (mostly `svg`s), I had to revert back to the old matcher and for now live with the fact that that will trigger the `Error Boundary`.

**Comment by ChiuMungZitAlexander** [comment]
Got this error by using `src` path. Solved by using `app` path. 👀👀

**Reply by dersimoezdag** [reply]
Can you provide some details? I don't know what exactly you've changed

**Reply by ChiuMungZitAlexander** [reply]
```
import createMiddleware from "next-intl/middleware";

export default createMiddleware({
  // A list of all locales that are supported
  locales: ["en", "zh"],

  // If this locale is matched, pathnames work without a prefix (e.g. `/about`)
  defaultLocale: "en",
});

export const config = {
  // Skip all paths that should not be internationalized. This example skips
  // certain folders and all pathnames with a dot (e.g. favicon.ico)
  matcher: ["/((?!api|_next|_vercel|.*\\..*).*)"],
};
```

<img width="397" alt="image" src="https://github.com/amannn/next-intl/assets/22234622/73772842-1610-4f82-b60c-d6c1930e9a66">

Middleware and file structure above works for me.

**Reply by dersimoezdag** [reply]
I get the same error with next@13.5.1 and 3.0.0-beta.19

**Comment by MrSchnitz** [comment]
For me the issue was that I had `messages` folder called `translations` instead

**Comment by IRediTOTO** [comment]
F* this error, it made me all days to fix just for translation :) Now I uninstall package

**Comment by amannn** [comment]
There was an update to this in https://github.com/amannn/next-intl/pull/742: `next-intl` will now call `notFound()` for these cases and therefore no error will be thrown anymore. A dev-only warning will still be printed though. Please still make sure that your middleware matcher is set up correctly. Have a look at the PR for details!

**Reply by MichalMoravik** [reply]
This is great, works like a charm! Thank you so much!

I suggest other people to check the `middleware.ts` in the docs examples. Specifically [this](https://github.com/amannn/next-intl/blob/main/examples/example-app-router/src/middleware.ts) was changed one week ago and served as a great example for my case.

**Comment by CarrettaRiccardo** [comment]
For me the problem was that I was using `withHydrationOverlay` in `next.config.mjs` from the package `@builder.io/react-hydration-overlay`, and somehow that interferes with next-intl

**Reply by AP-Forever** [reply]
How did you solve it?

**Comment by onurays** [comment]
For us the problem was: it was working on my computer but not on my colleague's. And we figured out that .env.local file is not there 🤦‍♂️ which was keeping supabase keys.

**Comment by AugustineSena** [comment]
Here is my ```middleware.ts``` that seems work for me but I don't know about security concern. I'm included also the ```img``` folder since almost all of my images are from there. The ```matcher ``` is modified value from this [StackOverflow](https://stackoverflow.com/questions/77508398/error-unable-to-find-next-intl-locale-because-the-middleware-didnt-run-on-th)

```
import createMiddleware from 'next-intl/middleware';

export default createMiddleware({
    // A list of all messages that are supported
    locales: ['en', 'id'],

    // Used when no locale matches
    defaultLocale: 'en'
});

export const config = {
    // Match only internationalized pathnames
    matcher: [
        '/((?!api|_next/static|_next/image|favicon.ico|apple-touch-icon.png|favicon.svg|img/).*)',
        '/',
        '/(id|en)/:path*',
    ]
};
```
